FearlessUtils/Classes/Crypto/BIP32KeyFactory.swift

```diff
@@ -0,0 +1,111 @@
+import CommonCrypto
+import IrohaCrypto
+import BigInt
+
+public enum BIP32KeyFactoryError: Error {
+    case invalidChildKey
+}
+
+protocol BIP32KeyFactoryProtocol {
+    func deriveFromSeed(_ seed: Data) throws -> BIP32ExtendedKeypair
+    func createKeypairFrom(_ parentKeypair: BIP32ExtendedKeypair, chaincode: Chaincode) throws -> BIP32ExtendedKeypair
+}
+
+public struct BIP32KeyFactory {
+    private let initialSeed = "Bitcoin seed"
+    let internalFactory = SECKeyFactory()
+
+    private func generateHMAC512(
+        from originalData: Data,
+        secretKeyData: Data
+    ) throws -> Data {
+        let digestLength = Int(CC_SHA512_DIGEST_LENGTH)
+        let algorithm = CCHmacAlgorithm(kCCHmacAlgSHA512)
+
+        var buffer = [UInt8](repeating: 0, count: digestLength)
+
+        originalData.withUnsafeBytes {
+            let rawOriginalDataPtr = $0.baseAddress!
+
+            secretKeyData.withUnsafeBytes {
+                let rawSecretKeyPtr = $0.baseAddress!
+
+                CCHmac(
+                    algorithm,
+                    rawSecretKeyPtr,
+                    secretKeyData.count,
+                    rawOriginalDataPtr,
+                    originalData.count,
+                    &buffer
+                )
+            }
+        }
+
+        return Data(bytes: buffer, count: digestLength)
+    }
+}
+
+extension BIP32KeyFactory: BIP32KeyFactoryProtocol {
+    func deriveFromSeed(_ seed: Data) throws -> BIP32ExtendedKeypair {
+        let hmacResult = try generateHMAC512(
+            from: seed,
+            secretKeyData: Data(initialSeed.utf8)
+        )
+
+        let privateKey = try SECPrivateKey(rawData: hmacResult[...31])
+        let chainCode = hmacResult[32...]
+
+        let keypair = try internalFactory.derive(fromPrivateKey: privateKey)
+
+        return BIP32ExtendedKeypair(keypair: keypair, chaincode: chainCode)
+    }
+
+    func createKeypairFrom(
+        _ parentKeypair: BIP32ExtendedKeypair,
+        chaincode: Chaincode
+    ) throws -> BIP32ExtendedKeypair {
+        let sourceData: Data = try {
+            switch chaincode.type {
+            case .hard:
+                let padding = try Data(hexString: "0x00")
```

Comment:
looks like overhead, consider to use `Data(repeating: 0 count: 1)` to remove an option where we can receive exception but it never happens.

Reply:
Done